examples/gguf-split/gguf-split.cpp

```diff
@@ -0,0 +1,510 @@
+        }
+        if (arg == "--split-tensors-size") {
+            if (++arg_idx >= argc) {
+                invalid_param = true;
+                break;
+            }
+            arg_found = true;
+            params.n_split_tensors = atoi(argv[arg_idx]);
+        }
+        if (arg == "--upload") {
+            arg_found = true;
+            params.operation = SPLIT_OP_UPLOAD;
+        }
+
+        if (!arg_found) {
+            throw std::invalid_argument("error: unknown argument: " + arg);
+        }
+    }
+
+    if (invalid_param) {
+        throw std::invalid_argument("error: invalid parameter for argument: " + arg);
+    }
+
+    if (argc - arg_idx < 2) {
+        printf("%s: bad arguments\n", argv[0]);
+        split_print_usage(argv[0]);
+        return false;
+    }
+
+    params.input = argv[arg_idx++];
+    params.output = argv[arg_idx++];
+
+    return true;
+}
+
+static bool split_params_parse(int argc, const char ** argv, split_params & params) {
+    bool result = true;
+    try {
+        if (!split_params_parse_ex(argc, argv, params)) {
+            split_print_usage(argv[0]);
+            exit(1);
+        }
+    }
+    catch (const std::invalid_argument & ex) {
+        fprintf(stderr, "%s\n", ex.what());
+        split_print_usage(argv[0]);
+        exit(1);
+    }
+    return result;
+}
+
+static void zeros(std::ofstream & file, size_t n) {
+    char zero = 0;
+    for (size_t i = 0; i < n; ++i) {
+        file.write(&zero, 1);
+    }
+}
+
+static std::string split_file_name(const std::string & path, int i_split, int n_split) {
+    char f_split[SPLIT_FILENAME_MAX] = {0};
+    snprintf(f_split, sizeof(f_split), SPLIT_FILENAME_FORMAT, path.c_str(), i_split + 1, n_split);
+    return std::string(f_split);
+}
+
+struct split_strategy {
+
+  const split_params params;
+  std::ifstream & f_input;
+  struct gguf_context * ctx_gguf;
+  struct ggml_context * ctx_meta = NULL;
+  const int n_tensors;
+
+  const int n_split;
+  int i_split = 0;
+
+  int i_tensor = 0;
+
+  std::vector<uint8_t> read_data;
+
+  struct gguf_context * ctx_out;
+  std::ofstream fout;
+
+  split_strategy(const split_params & params,
+                 std::ifstream & f_input,
+                 struct gguf_context * ctx_gguf,
+                 struct ggml_context * ctx_meta) :
+      params(params),
+      f_input(f_input),
+      ctx_gguf(ctx_gguf),
+      ctx_meta(ctx_meta),
+      n_tensors(gguf_get_n_tensors(ctx_gguf)),
+      n_split(std::ceil(1. * n_tensors / params.n_split_tensors)) {
+  }
+
+  bool should_split() const {
+      return i_tensor < n_tensors && i_tensor % params.n_split_tensors == 0;
+  }
+
+  void split_start() {
+      ctx_out = gguf_init_empty();
+
+      // Save all metadata in first split only
+      if (i_split == 0) {
+          gguf_set_kv(ctx_out, ctx_gguf);
+          gguf_set_val_u8(ctx_out, LLM_KV_GENERAL_SPLIT_N_SPLIT, n_split);
+      }
+      gguf_set_val_u8(ctx_out, LLM_KV_GENERAL_SPLIT_I_SPLIT, i_split);
```

Comment:
Should we also have `LLM_KV_GENERAL_SPLIT_N_SPLIT` inside each split? Maybe redundant, but maybe more consistent if `general.split` and `general.split_count` always come together